src/wallet/wallet.cpp

```diff
@@ -3679,38 +3679,42 @@ void CWallet::GetKeyBirthTimes(std::map<CKeyID, int64_t>& mapKeyBirth) const {
  * https://bitcointalk.org/?topic=54527, or
  * https://github.com/bitcoin/bitcoin/pull/1393.
  */
-unsigned int CWallet::ComputeTimeSmart(const CWalletTx& wtx) const
+unsigned int CWallet::ComputeTimeSmart(const CWalletTx& wtx, bool rescanning_old_block) const
 {
     unsigned int nTimeSmart = wtx.nTimeReceived;
     if (!wtx.isUnconfirmed() && !wtx.isAbandoned()) {
         int64_t blocktime;
         if (chain().findBlock(wtx.m_confirm.hashBlock, FoundBlock().time(blocktime))) {
-            int64_t latestNow = wtx.nTimeReceived;
-            int64_t latestEntry = 0;
-
-            // Tolerate times up to the last timestamp in the wallet not more than 5 minutes into the future
-            int64_t latestTolerated = latestNow + 300;
-            const TxItems& txOrdered = wtxOrdered;
-            for (auto it = txOrdered.rbegin(); it != txOrdered.rend(); ++it) {
-                CWalletTx* const pwtx = it->second;
-                if (pwtx == &wtx) {
-                    continue;
-                }
-                int64_t nSmartTime;
-                nSmartTime = pwtx->nTimeSmart;
-                if (!nSmartTime) {
-                    nSmartTime = pwtx->nTimeReceived;
-                }
-                if (nSmartTime <= latestTolerated) {
-                    latestEntry = nSmartTime;
-                    if (nSmartTime > latestNow) {
-                        latestNow = nSmartTime;
+            if(!rescanning_old_block) {
```

Comment:
commit 4997a2ea5f2 "wallet: fix ComputeTimeSmart algorithm to use blocktime during old block rescanning fixes #20181"

- does the documentation need to be updated?

- missing space
```suggestion
            if (!rescanning_old_block) {
```

- maybe start with the truthy case first, e.g. `if (rescanning_old_block) {`...

- can drop the "fixes #20181" from the commit message; in the PR description is fine (it will be copied by the merge commit)

- suggestion to reviewers: review this commit diff with `-w`

Reply:
done locally, except doxygen documentation update, not pushed yet

Reply:
I'm not sure about how to update the doxygen documentation, because I don't understand the way the logic is explained. The documentation split sending and receiving transaction treatment explanation, but there's no distinction in ComputeTimeSmart in the code (or I don't see it).

That's why I'm not sure between which lines I should insert the new beahaviour description.
Maybe something like "- If receiving a transaction during a rescanning process, assign all its (not already known) transactions' timestamps to the block time." between the second and the third one :

```
/**
 * Compute smart timestamp for a transaction being added to the wallet.
 *
 * Logic:
 * - If sending a transaction, assign its timestamp to the current time.
 * - If receiving a transaction outside a block, assign its timestamp to the
 *   current time.
 * - If receiving a transaction during a rescanning process, assign all its
 *   (not already known) transactions' timestamps to the block time.
 * - If receiving a block with a future timestamp, assign all its (not already
 *   known) transactions' timestamps to the current time.
 * - If receiving a block with a past timestamp, before the most recent known
 *   transaction (that we care about), assign all its (not already known)
 *   transactions' timestamps to the same timestamp as that most-recent-known
 *   transaction.
 * - If receiving a block with a past timestamp, but after the most recent known
 *   transaction, assign all its (not already known) transactions' timestamps to
 *   the block time.
 *
 * For more information see CWalletTx::nTimeSmart,
 * https://bitcointalk.org/?topic=54527, or
 * https://github.com/bitcoin/bitcoin/pull/1393.
 */
```

Reply:
I've push the proposition in a dedicated commit and will revise it if needed.